master/internal/restore.go

```diff
@@ -134,79 +133,68 @@ func (m *Master) restoreExperiment(expModel *model.Experiment) error {
 	return nil
 }
 
-// restoreTrial takes the a searcher.Create and attempts to restore the trial that would be
-// associated with it. On failure, the trial is just reset to the start and errors are logged.
 func (e *internalExperiment) restoreTrial(
 	ckpt *model.Checkpoint, searcher experiment.TrialSearcherState,
 ) {
-	l := e.syslog.WithField("request-id", searcher.Create.RequestID)
-	l.Debug("restoring trial")
+	syslog := e.syslog
+	syslog.Debug("restoring trial")
 
+	var trial *model.Trial
 	var trialID *int
-	var terminal bool
-	switch trial, err := db.TrialByExperimentAndRequestID(context.TODO(),
-		e.ID, searcher.Create.RequestID); {
-	case errors.Cause(err) == db.ErrNotFound:
-		l.Debug("trial was not previously persisted")
-	case err != nil:
-		// This is the only place we _have_ to error, because if the trial did previously exist
-		// and we failed to retrieve it, continuing will result in an invalid state (we'll get a
-		// new trial in the trials table with the same (experiment_id, request_id).
-		l.WithError(err).Error("failed to retrieve trial, aborting restore")
-		terminal = true
-	default:
-		trialID = &trial.ID
-		l = l.WithField("trial-id", trial.ID)
-		if model.TerminalStates[trial.State] {
-			l.Debugf("trial was in terminal state in restore: %s", trial.State)
-			terminal = true
+	var err error
+
+	if searcher.TrialID != nil {
+		if _, ok := e.trials[*searcher.TrialID]; ok {
+			syslog.Errorf("trial %d was already restored, exiting", *searcher.TrialID)
+			return
+		}
+		trial, err = db.TrialByID(context.TODO(), int(*searcher.TrialID))
+		if err != nil {
+			syslog.WithError(err).Error("failed to retrieve previous trial, restoring new trial")
+		} else {
+			syslog = syslog.WithField("trial-id", trial.ID)
 		}
 	}
 
-	taskID := trialTaskID(e.ID, searcher.Create.RequestID)
-	if !terminal && trialID != nil {
-		trialTaskIDs, err := db.TrialTaskIDsByTrialID(context.TODO(), *trialID)
+	taskID := model.TaskID(fmt.Sprintf("%d.%s", e.ID, model.NewTaskID()))
+
+	if trial != nil {
+		trialID = &trial.ID
+		// If the run being restored was in a terminal state, replay the close for the searcher and exit.
+		if model.TerminalStates[trial.State] {
+			syslog.Debugf("trial was in terminal state in restore: %s", trial.State)
+			if !e.searcher.TrialIsClosed(*searcher.TrialID) {
+				e.trialExited(*searcher.TrialID, nil)
+			}
+			return
+		}
+		trialTaskIDs, err := db.TrialTaskIDsByTrialID(context.TODO(), trial.ID)
 		switch {
 		case err != nil:
-			l.WithError(err).Error("failed to retrieve trial's tasks, stopping restore")
-			terminal = true
+			syslog.WithError(err).Error("failed to retrieve tasks for run")
```

Comment:
we removed the `terminal` bit. i think we need to add it back and decide to close the searcher in some of these cases where we don't have enough information to continue.